test/integration/buildrun_cleanup_test.go

```diff
@@ -0,0 +1,454 @@
+		})
+
+		It("Should not find the buildrun few seconds after it succeeds", func() {
+
+			Expect(tb.CreateBuild(buildObject)).To(BeNil())
+
+			buildObject, err = tb.GetBuildTillValidation(buildObject.Name)
+			Expect(err).To(BeNil())
+
+			// Make the TTL 5 minutes
+			buildRunObject.Spec.Retention.TTLAfterFailed.Duration = time.Minute * 5
+			Expect(tb.CreateBR(buildRunObject)).To(BeNil())
+			br, err := tb.GetBRTillCompletion(buildRunObject.Name)
+			Expect(err).To(BeNil())
+			Expect(br.Status.GetCondition(v1alpha1.Succeeded).Status).To(Equal(corev1.ConditionFalse))
+
+			// Make the TTL 5 seconds
+			br.Spec.Retention.TTLAfterFailed.Duration = time.Second * 5
+			Expect(tb.UpdateBR(br)).To(BeNil())
+			_, err = tb.GetBRTillNotFound(buildRunObject.Name, time.Second*1, time.Second*15)
+			Expect(apierrors.IsNotFound(err)).To(BeTrue())
+		})
+	})
+
+	Context("When a buildrun with short ttl set fails. TTL is defined in the buildrun specs", func() {
+
+		BeforeEach(func() {
+			buildSample = []byte(test.MinimalBuildWithRetentionLimitOne)
+			buildRunSample = []byte(test.MinimalBuildRunRetentionTTLFive)
+		})
+
+		It("Should not find the buildrun few seconds after it fails", func() {
+
+			Expect(tb.CreateBuild(buildObject)).To(BeNil())
+
+			buildObject, err = tb.GetBuildTillValidation(buildObject.Name)
+			Expect(err).To(BeNil())
+			Expect(tb.CreateBR(buildRunObject)).To(BeNil())
+
+			br, err := tb.GetBRTillCompletion(buildRunObject.Name)
+			Expect(err).To(BeNil())
+			Expect(br.Status.GetCondition(v1alpha1.Succeeded).Status).To(Equal(corev1.ConditionFalse))
+			_, err = tb.GetBRTillNotFound(buildRunObject.Name, time.Second*1, time.Second*15)
+			Expect(apierrors.IsNotFound(err)).To(BeTrue())
+		})
+	})
+
+	Context("When a buildrun that has TTL defined in its spec and on the corresponding build's spec fails", func() {
+
+		BeforeEach(func() {
+			buildSample = []byte(test.MinimalBuildWithRetentionTTLOneMin)
+			buildRunSample = []byte(test.MinimalBuildRunRetentionTTLFive)
+		})
+
+		It("Should honor the TTL defined in the  buildrun.", func() {
+			Expect(tb.CreateBuild(buildObject)).To(BeNil())
+
+			buildObject, err = tb.GetBuildTillValidation(buildObject.Name)
+			Expect(err).To(BeNil())
+			Expect(tb.CreateBR(buildRunObject)).To(BeNil())
+
+			br, err := tb.GetBRTillCompletion(buildRunObject.Name)
+			Expect(err).To(BeNil())
+			Expect(br.Status.GetCondition(v1alpha1.Succeeded).Status).To(Equal(corev1.ConditionFalse))
+			_, err = tb.GetBRTillNotFound(buildRunObject.Name, time.Second*1, time.Second*15)
+			Expect(apierrors.IsNotFound(err)).To(BeTrue())
+
+		})
+	})
+
+	Context("Multiple failed buildruns related to build with limit 1", func() {
+
+		BeforeEach(func() {
+			buildSample = []byte(test.MinimalBuildWithRetentionLimitOne)
+			buildRunSample = []byte(test.MinimalBuildRunRetention)
+		})
+
+		It("Should not find the older failed buildrun", func() {
+			buildObject.Spec.Source.ContextDir = nil
```

Comment:
Is this still needed with the param value set to false ?

```suggestion
			buildObject.Spec.Source.ContextDir = nil
```